modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -691,40 +687,103 @@ public String getKey() {
+        final Map<StatsKey, OperationMetrics> collectors = new ConcurrentHashMap<>();
 
         Map<String, Long> statsMap(boolean stateless) {
             if (stateless) {
                 return collectors.entrySet()
                     .stream()
-                    .collect(Collectors.toUnmodifiableMap(e -> e.getKey().toString(), e -> e.getValue().sum()));
+                    .collect(Collectors.toUnmodifiableMap(e -> e.getKey().toString(), e -> e.getValue().count()));
             } else {
                 Map<String, Long> normalisedStats = Arrays.stream(Operation.values()).collect(Collectors.toMap(Operation::getKey, o -> 0L));
                 collectors.forEach(
                     (key, value) -> normalisedStats.compute(
                         key.operation.getKey(),
-                        (k, current) -> Objects.requireNonNull(current) + value.sum()
+                        (k, current) -> Objects.requireNonNull(current) + value.count()
                     )
                 );
                 return Map.copyOf(normalisedStats);
             }
         }
 
-        public void onSuccessfulRequest(Operation operation, OperationPurpose purpose) {
-            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new LongAdder()).increment();
+        public void onRequestComplete(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
+            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new OperationMetrics(operation, purpose))
+                .onRequestComplete(requestMetrics);
         }
     }
 
+    // visible for testing
+    class OperationMetrics {
+
+        final LongAdder counter;
+        final Map<String, Object> attributes;
+
+        private OperationMetrics(Operation operation, OperationPurpose purpose) {
+            this.counter = new LongAdder();
+            this.attributes = RepositoriesMetrics.createAttributesMap(repositoryMetadata, purpose, operation.getKey());
+        }
+
+        public void onRequestComplete(AzureClientProvider.RequestMetrics requestMetrics) {
+            counter.add(requestMetrics.getRequestCount());
+
+            // range not satisfied is not retried, so we count them by checking the final response
+            if (requestMetrics.getStatusCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
+                repositoriesMetrics.requestRangeNotSatisfiedExceptionCounter().incrementBy(1, attributes);
+            }
+
+            repositoriesMetrics.operationCounter().incrementBy(1, attributes);
+            if (requestMetrics.getStatusCode() <= 199 || requestMetrics.getStatusCode() > 299) {
+                repositoriesMetrics.unsuccessfulOperationCounter().incrementBy(1, attributes);
+            }
+
+            repositoriesMetrics.requestCounter().incrementBy(requestMetrics.getRequestCount(), attributes);
+            if (requestMetrics.getErrorCount() > 0) {
```

Comment:
I wonder if we're losing information by not sampling when the error/throttling count is zero. This is consistent with the S3 metrics but I feel it makes it hard to distinguish between an increase in the error rate and an increase in the volume of requests?

Reply:
Maybe. So far it has been relatively easy to differentiate by just eyeballing the `requests` and `exceptions` lines on the visualization. Part of the reason maybe the increase of requests is currently dominated by internal schedulings, such as SLM runs. So they have been quite obvious. When we have more larger projects that bring in more bursty loads, we might need to introduce something like ratio for more precise differentiation. I think it can wait for now.